### PR TITLE
DM-37744: Make daf_butler compatible with sqlalchemy 2

### PR DESCRIPTION
**Pull request by andy-slac**
Lots of changes everywhere to make it run with sqlalchemy 2. Many mypy fixes. I enabled sqlalchemy 2 in requirements, so GA should build with that.

## Checklist

- [X] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/daf/butler/registry/interfaces/_database.py [hunk 1]
```diff
@@ -1763,7 +1774,9 @@ def query(
             connection = self._engine.connect()
         else:
             connection = self._session_connection
-        result = connection.execute(sql, *args, **kwargs)
+        # TODO: SelectBase is not good for execute(), but it used everywhere,
+        # e.g. in daf_relation. We should switch to Executable at some point.
```
Comment (timj) [inline python/lsst/daf/butler/registry/interfaces/_database.py:1778]: Can you make a ticket for this if daf_relation is using some old syntax? cc/ @TallJimbo

File: python/lsst/daf/butler/core/ddl.py [hunk 2]
```diff
@@ -249,9 +251,14 @@ def process_bind_param(self, value: Any, dialect: sqlalchemy.Dialect) -> Optiona
...
             return "%.32x" % value.int
 
-    def process_result_value(self, value: Optional[str], dialect: sqlalchemy.Dialect) -> Optional[uuid.UUID]:
+    def process_result_value(
+        self, value: str | uuid.UUID | None, dialect: sqlalchemy.Dialect
+    ) -> Optional[uuid.UUID]:
         if value is None:
             return value
+        elif isinstance(value, uuid.UUID):
+            # sqlalchemy 2 converts to UUID internally
```
Comment (timj) [inline python/lsst/daf/butler/core/ddl.py:260]: Does that mean in sqlalchemy 1.4 we can't do this?

File: python/lsst/daf/butler/registry/databases/postgresql.py [hunk 3]
```diff
@@ -197,7 +200,7 @@ def _convertExclusionConstraintSpec(
         metadata: sqlalchemy.MetaData,
     ) -> sqlalchemy.schema.Constraint:
         # Docstring inherited.
-        args = []
+        args: list[tuple[sqlalchemy.schema.Column, str]] = []
```
Comment (timj) [inline python/lsst/daf/butler/registry/databases/postgresql.py:203]: Are you surprised this method never gets called?

File: python/lsst/daf/butler/registry/interfaces/_database.py [hunk 4]
```diff
@@ -57,7 +58,9 @@
 from .._exceptions import ConflictingDefinitionError
 
 
-def _checkExistingTableDefinition(name: str, spec: ddl.TableSpec, inspection: List[Dict[str, Any]]) -> None:
+# TODO: method is called with list[ReflectedColumn] in SA 2, and
```
Comment (timj) [inline python/lsst/daf/butler/registry/interfaces/_database.py:61]: So when we get SA2 we can make the type annotation explicit about that list contents?

**Review by timj — APPROVED** [review]
Thanks. Looks great. I can confirm that the tests no longer have any sqlalchemy warnings (and I believe we still have that env var turned on)

File: python/lsst/daf/butler/registry/databases/postgresql.py (hunk 3 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/databases/postgresql.py:203]: I'm surprised.  I'd expect this to be called whenever we register a calibration dataset type with a new set of dimensions.

File: python/lsst/daf/butler/registry/interfaces/_database.py (hunk 1 above)
Comment (andy-slac) [reply python/lsst/daf/butler/registry/interfaces/_database.py:1778]: I made DM-37971 for @TallJimbo, but you can reassign it to me if you don't have time to work on that.

File: python/lsst/daf/butler/core/ddl.py (hunk 2 above)
Comment (andy-slac) [reply python/lsst/daf/butler/core/ddl.py:260]: In 1.4 strings are returned for UUIDs for both Postgres and sqlite, strings are still handled in `else` clause below.

File: python/lsst/daf/butler/registry/databases/postgresql.py (hunk 3 above)
Comment (andy-slac) [reply python/lsst/daf/butler/registry/databases/postgresql.py:203]: Grepping through the code it looks like that it is only called when `TableSpec.exclusion` is not empty, and that could only happen when timespan representation class `hasExclusionConstraint()` returns `True`. I found only one implementation of `hasExclusionConstraint()` which always returns `False`.

File: python/lsst/daf/butler/registry/interfaces/_database.py (hunk 4 above)
Comment (andy-slac) [reply python/lsst/daf/butler/registry/interfaces/_database.py:61]: Right,  we should clean it up when we decide to drop 1.4 support.

File: python/lsst/daf/butler/registry/databases/postgresql.py (hunk 3 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/databases/postgresql.py:203]: Oh, interesting.  The PostgreSQL-specific timespan representation should have `hasExclusionConstraint()` return `True`, as that's one of the things that `btree_gist` brings, and would avoid a table lock (IIRC) in `certify` (which I thought we were already avoiding for PostgreSQL).

But I'm not sure if we can just fix that without a migration now that we've made the mistake.

**Pull request merged**